docs/langref/relay_expr.rst

```diff
@@ -0,0 +1,552 @@
+    // %f is a closure where %x maps to Constant(0, (10, 10), float32)
+    let %f = %g();
+    let %x = Constant(1, (10, 10), float32);
+    %f(%x) // evaluates to `Constant(0, (10, 10), float32)
+
+Polymorphism and Type Relations
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+.. *Note: type parameter syntax is not yet supported in the text format.*
+
+A function may also be given a set of type parameters, which can be
+substituted for specific types at call sites. Functions with
+type parameters are *type polymorphic*; their return type or the types
+of arguments they will accept can vary based on the type arguments
+given at call sites.
+
+For example, one can define a polymorphic identity function for
+any Relay type as follows:
+
+.. code-block:: python
+    fn<t: Type>(%x : t) -> t {
+        %x
+    }
+
+The below definition is also polymorphic, but restricts its
+arguments to tensor types:
+
+.. code-block:: python
+    fn<s: Shape, bt: BaseType>(%x: Tensor[s, bt]) {
+        %x
+    }
+
+Notice that the return type is omitted and will be inferred.
+
+.. *Note: :code:`where` syntax is not yet supported in the text format.*
+
+A function may also be subject to one or more type relations, such as in
+the following:
+
+.. code-block:: python
+
+    fn(%x, %y) where Broadcast { add(%x, %y) }
+
+In the above definition, the types of :code:`%x` and :code:`%y` and the return type
+are subject to the :code:`Broadcast` relation, meaning all three must be tensors
+and their shapes follow the elementwise broadcast relation. As with
+operators, the definitions of relations are not transparent to Relay
+and they are instead implemented externally in either C++ or Python.
+
+As in the case of :code:`Broadcast`, relations are used to express complicated
+constraints on types (especially tensor shapes).
+All function relations must hold at all call sites;
+type checking is thus treated as a constraint-solving problem.
+For more detail on type relations and their implementations,
+please see the documentation on `Relay's Type System`_.
+
+Operators
+=========
+
+An operator is a primitive operation, such as :code:`add` or :code:`conv2d`, not defined in the Relay
+language. Operators are declared in the global operator
+registry in C++. Many common operators are backed by TVM's
+Tensor Operator Inventory (`TOPI <https://github.com/dmlc/tvm/tree/master/topi>`__).
+
+To register an operator a user must provide an implementation
+of the operator, its type, and any other desired metadata.
+The operator registry is a column-based store where
+operators are keys, so any metadata (which might be referenced
+by optimization passes) may be registered as a new column.
+
+From the perspective of Relay's type system, an operator is a function,
+so operators may be called like any other function and have function
+types. In particular, operator types are registered using a single
+type relation (see `Polymorphism and Type Relations`_), typically a relation
```

Comment:
it might be useful to reference the type system discussion of type relations here, because it's talking about implementing type relations

Reply:
Perhaps I will, but I think the linked subsection does reference that.